### PR TITLE
Implement download feature

### PR DESCRIPTION
**Pull request by michaelbretagne**
* Add colors to the svg download icons (red when download is not ready, green when it is ready for download).
* Allow users to download the visualization when the song finishes (Download button turn green and it is clickable for download).
* Activate play button (visualization is shown on the screen and the canvas/audio is recorded)
* Fix play/pause button behavior (button pause return to button play when the song end)
* Display duration of the song on the play bar
* Refactor App.js class component to a react hook functional component
* Fix some CSS so the canvas is shown correctly.

### CODE REVIEW

File: src/components/Visualizer/Visualizer.module.scss
Comment (ArnasDickus) [inline src/components/Visualizer/Visualizer.module.scss]: Why did you changed these css values?

**Review by ArnasDickus — CHANGES_REQUESTED** [review]
Hello thanks for adding PR.

From user perspective a few issues:
1) The only way for user to download the song visuals is to listen and wait till the song is fully finished. Is there a way to just upload the song and computer could somehow record visuals behind the scenes?
2) It's just me or visuals are kind of slow/lagging.
3) Downloaded files using vlc player, won't display time, and user on vlc player can't skip forward.

File: src/components/Visualizer/Visualizer.module.scss
Comment (michaelbretagne) [reply src/components/Visualizer/Visualizer.module.scss]: I just applied a quick fix that Askat suggested on Discord. I did not check if those values where necessary or not.
I can remove all the CSS styles I made and someone else can properly fix the CSS issues in another PR.

**Comment by michaelbretagne** [comment]
For the issues:
1. Unfortunately, I think it is the only way to achieve the recording. It is recording live on what the canvas is showing. It is like taking a video of the computer screen using your smartphone. If nothing is shown, nothing can be recorded. I tried different things but I could not find any other solution. Maybe someone will find another way but it is all I got for now. 
2. Yes, there is maybe a bit of lag. The frame rate of the recording can be changed easily. I will increase the frame rate value and it should be better.
 3. I look around but I did not find any solution for that at the moment.

**Comment by michaelbretagne** [comment]
**Changes made:**

**In AppBrowser.js:** 
- Remove downloadState props selector (no need to get this props from redux in this component)
- Make sure we are redirected to App.js only when the song is fetched from firebase

**In App.js:** 
- Add downloadState useSelector to get the props from redux

**In PlayerBar.js:** 
- Make sure if the user press the play button, the download button is not clickable and it turns red.

**In Sketch.js:** 
- Increase the recording frame rate so the downloaded file looks a bit better.
- Fix the display time of the downloaded files so the users can skip forward/backward.
- Add the possibility to resume a recording after the user press pause. Can see it micro pause of the visual on the downloaded file. Not sure what we can do about it.
- Make sure the visualization starts when the user press play. Before if the user pressed play too fast after we are directed to App.js the visualization was not starting. Now it will start as soon as the song is loaded in p5 (only if the play button is pressed, of course).
- If the song ends and the user plays the song again a new recording is created.

**In Visualizer.module.scss:** 
- Remove change I made in the last commit  (Need to create another task to fix how the navbar/visualizer/logo are displayed on the screen)

I will try to think about another solution to record the visualization without having to wait for the end of the song.

**Review by ArnasDickus — CHANGES_REQUESTED** [review]
Thanks for adding PR. 
`Make sure we are redirected to App.js only when the song is fetched from firebase.`
I think we need to add some kind of spinner in the center after uploading to inform the user that the application is not frozen.

`Make sure if the user press the play button, the download button is not clickable and it turns red.`
Controversial decision, from discussion in Visual Music Core Team. Yogiha position, is that if user want's he should be able to download not finished song visuals. However we could change this later, so I will approve for now. So you can ignore this.

`Increase the recording frame rate so the downloaded file looks a bit better.`
Yes it does look better. although there is some issues. For example using VLC media player, it won't allow to move forward recording. In windows 10 `movies & Tv` main application. It display only sound. So it might be wierd conversion or file format not fully supported.

`Fix the display time of the downloaded files so the users can skip forward/backward.` Can't confirm this with vlc media player. I still can't skip forward, could you sent me a working example?

`Add the possibility to resume a recording after the user press pause. Can see it micro pause of the visual on the downloaded file. Not sure what we can do about it.`
Yes this can become a common problem, users are impatient and pausing, and visualizer is getting distorted. I'am also stumped.

**Comment by michaelbretagne** [comment]
I added a spinner. Not sure if it fits the style. Not really related to the download feature anyway.
The video downloaded work as expected when using Firefox.
Apparently there is a well-known bug in Chrome. Basically, the duration of the recorded media isn't added to the headers of the final file.
Sadly, this bug is currently marked as WontFix by the Chromium team.
There is some workaround that works for some people. It did not work for me.
So I am not sure what to do about that.
Also, I am using a Macbook pro to code. I don't have Windows 10 so I am not sure how to try/test/fix anything related to Microsoft product.

**Comment by michaelbretagne** [comment]
I made some modifications based on YogiHa suggestions.

I moved the download logic into his own component. I kept the button style into the PlayerBar.module.scss. I thought it would make more sense to keep it there.

Also, as YogiHa pointed out, the App component is better to be exported as default since it is a component. So I changed that. Good catch YogiHa!

In AppBrowser, instead of checking for {song && song.url} I modified the InitialProps in songReducer.js so we can just check for {song.url} before to redirect to App.js.

I also added a dispatch reference to Visualizer and passed it as a prop to Sketch.js. 
I used it to dispatch downloadVisualEnd() action after the download happened instead of using a setTimeout to fire downloadVisualEnd().
This dispatch reference will also be helpful later when we will need to dispatch other actions (like song.currentTime) from Sketch.js. Thx to YogiHa to show me that we could do that.

**Review by ArnasDickus — APPROVED** [review]
I have no more issues. One more thing, I suggest instead of working directly to the development branch, from the development branch create unique branch and work from it instead. For a few reasons:
1)  Until I approve of this PR, you can't work on another task.
2) Sometimes when working directly on the development branch you get other contributors PR. So I just to just create another unique branch.

**Pull request merged**
